File: src/utils/decode_util.rs
/// Decode winget's stdout bytes to a String.
/// Handles UTF-16 LE BOM, UTF-8 BOM, and plain UTF-8.
pub fn decode_winget_output(bytes: &[u8]) -> String {
    if bytes.starts_with(&[0xFF, 0xFE]) {
        let words: Vec<u16> = bytes[2..]
            .chunks_exact(2)
            .map(|b| u16::from_le_bytes([b[0], b[1]]))
            .collect();
        return String::from_utf16_lossy(&words);
    }
    if bytes.starts_with(&[0xEF, 0xBB, 0xBF]) {
        return String::from_utf8_lossy(&bytes[3..]).into_owned();
    }
    String::from_utf8_lossy(bytes).into_owned()
}

/// Returns true for lines that should be hidden in the install log.
///
/// Winget uses bare \r to animate a progress spinner and download bar in
/// place. When stdout is piped those \r sequences arrive inside a single
/// BufReader "line". We split on \r here and filter out the noise segments
/// so only meaningful status lines reach the UI.
pub fn is_noise_line(s: &str) -> bool {
    let t = s.trim();
    if t.is_empty() { return true; }
    if t == "-" || t == "\\" || t == "|" || t == "/" { return true; }
    if t.contains('█') || t.contains('▒') { return true; }
    if t.contains("KB /") || t.contains("MB /") || t.contains("GB /") { return true; }
    if t.ends_with('%') && t.trim_end_matches('%').trim().chars().all(|c| c.is_ascii_digit()) {
        return true;
    }
    false
}

File: Dotfile-java-TUI/src/main/java/io/github/sampongstarluck/dotfile/service/DecodeUtil.java
package io.github.sampongstarluck.dotfile.service;

import java.nio.charset.StandardCharsets;

/** Port of {@code ../src/utils/decode_util.rs}. */
public final class DecodeUtil {

    private DecodeUtil() {}

    /** winget may emit UTF-16 LE with BOM; also handle UTF-8 BOM; else UTF-8. */
    public static String decodeWingetOutput(byte[] bytes) {
        if (bytes.length >= 2 && (bytes[0] & 0xFF) == 0xFF && (bytes[1] & 0xFF) == 0xFE) {
            return new String(bytes, 2, bytes.length - 2, StandardCharsets.UTF_16LE);
        }
        if (bytes.length >= 3 && (bytes[0] & 0xFF) == 0xEF && (bytes[1] & 0xFF) == 0xBB
                && (bytes[2] & 0xFF) == 0xBF) {
            return new String(bytes, 3, bytes.length - 3, StandardCharsets.UTF_8);
        }
        return new String(bytes, StandardCharsets.UTF_8);
    }

    /** Port of {@code is_noise_line()}: filter winget spinner/progress junk from install-log output. */
    public static boolean isNoiseLine(String s) {
        String t = s.trim();
        if (t.isEmpty()) return true;
        if (t.equals("-") || t.equals("\\") || t.equals("|") || t.equals("/")) return true;
        if (t.contains("█") || t.contains("▒")) return true;
        if (t.contains("KB /") || t.contains("MB /") || t.contains("GB /")) return true;
        if (t.endsWith("%")) {
            String num = t.substring(0, t.length() - 1).trim();
            if (!num.isEmpty() && num.chars().allMatch(Character::isDigit)) return true;
        }
        return false;
    }
}
